# mapping FASTQ files


This describes a general procedure not taking into account improvements/filtrations of the input FASTQ files such as:

* optical duplicates removal
* trimming reads containing any of the 3 relevant NEXTflex primers (both forward and reverse) which are detectable in a significant portion of reads


### genome

* version: GRCh38.p13 (used by GENCODE & ENSEMBL)
* source: xxx

Caveat: get the MT sequence

### tRNAs

There are two viable options:
1. 

